hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/hudi/cdc/CDCRelation.scala

```diff
@@ -0,0 +1,199 @@
+    endInstant: String,
+    options: Map[String, String]
+) extends BaseRelation with PrunedFilteredScan with Logging {
+
+  val spark: SparkSession = sqlContext.sparkSession
+
+  val (tableAvroSchema, _) = {
+    val schemaUtil = new TableSchemaResolver(metaClient)
+    val avroSchema = Try(schemaUtil.getTableAvroSchema) match {
+      case Success(schema) => schema
+      case Failure(e) =>
+        throw new IllegalArgumentException("Failed to fetch schema from the table", e)
+    }
+    // try to find internalSchema
+    val internalSchemaFromMeta = try {
+      schemaUtil.getTableInternalSchemaFromCommitMetadata.orElse(InternalSchema.getEmptyInternalSchema)
+    } catch {
+      case _: Exception => InternalSchema.getEmptyInternalSchema
+    }
+    (avroSchema, internalSchemaFromMeta)
+  }
+
+  val tableStructSchema: StructType = AvroConversionUtils.convertAvroSchemaToStructType(tableAvroSchema)
+
+  val cdcExtractor: HoodieCDCExtractor = new HoodieCDCExtractor(metaClient, startInstant, endInstant)
+
+  override final def needConversion: Boolean = false
+
+  override def schema: StructType = CDCRelation.FULL_CDC_SPARK_SCHEMA
+
+  override def buildScan(requiredColumns: Array[String], filters: Array[Filter]): RDD[Row] = {
+    val internalRows = buildScan0(requiredColumns, filters)
+    internalRows.asInstanceOf[RDD[Row]]
+  }
+
+  def buildScan0(requiredColumns: Array[String], filters: Array[Filter]): RDD[InternalRow] = {
+    val nameToField = schema.fields.map(f => f.name -> f).toMap
+    val requiredSchema = StructType(requiredColumns.map(nameToField))
+    val originTableSchema = HoodieTableSchema(tableStructSchema, tableAvroSchema.toString)
+    val parquetReader = HoodieDataSourceHelper.buildHoodieParquetReader(
+      sparkSession = spark,
+      dataSchema = tableStructSchema,
+      partitionSchema = StructType(Nil),
+      requiredSchema = tableStructSchema,
+      filters = Nil,
+      options = options,
+      hadoopConf = spark.sessionState.newHadoopConf()
+    )
+
+    val changes = cdcExtractor.extractCDCFileSplits().values().asScala.map { pairs =>
+      HoodieCDCFileGroupSplit(
+        pairs.asScala.map(pair => (pair.getLeft, pair.getRight)).sortBy(_._1).toArray
+      )
+    }
+    val cdcRdd = new HoodieCDCRDD(
+      spark,
+      metaClient,
+      parquetReader,
+      originTableSchema,
+      schema,
+      requiredSchema,
+      changes.toArray
+    )
+    cdcRdd.asInstanceOf[RDD[InternalRow]]
+  }
+}
+
+object CDCRelation {
+
+  val CDC_OPERATION_DELETE: UTF8String = UTF8String.fromString(DELETE.getValue)
+  val CDC_OPERATION_INSERT: UTF8String = UTF8String.fromString(INSERT.getValue)
+  val CDC_OPERATION_UPDATE: UTF8String = UTF8String.fromString(UPDATE.getValue)
+
+  /**
+   * CDC Schema For Spark.
+   * Also it's schema when `hoodie.table.cdc.supplemental.logging.mode` is `cdc_data_before_after`.
+   * Here we use the debezium format.
+   */
+  val FULL_CDC_SPARK_SCHEMA: StructType = {
+    StructType(
```

Comment:
Do we need to expose the schema as debezium for reader internal ? Why not reuse the field `_hoodie_operaton` which is a hoodie format.

Curious how downstream pipeline handle these records ? For SQL users, they declare the table schema, for example with fields (a, b, c, d), now you return RDD with schema of avro, how and when to deserialize them into (a, b, c ,d) then ?

Reply:
the returned RDD use the json string, not the avro format (that only used inside of hudi).
`op`, `ts_ms`, `before` and `after` are the fields given in debezium.